### DISCUSSION TITLE
Welcome to WIP-MultiPSK-keygen Discussions!

### DISCUSSION THREAD

**Discussion by blkph0x** [blkph0x/WIP-MultiPSK-keygen#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ghost** [comment]
Hope you are ok m8, I have created a totally unlocked version of 4.48. I am trying to create a form to edit the USER.COD file, so users can personalise their registerd to information. Needs to save 100byes only and needs to run a a standalone and not an installer type programme. On a steep learning curve. Sent link on you video channel but dont think it is showing to download and try. Files also on my GIT. Kat sends

**Reply by blkph0x** [reply]
Yes we could make the keygen a dll and then we can call the dll once we grab flow of execution check if user.cod is precent if it isn't load the dll and ask for username. Else if user.cod file is there skip loading Dll jump to the serial stuff join my discord of ya Wana chat more about it https://discord.gg/y9YzuRfM

**Reply by blkph0x** [reply]
Ohh got you I don't really play with VB that often I use mainly C C++ and asm but you should beable to deploy as standalone .exe or move to C++ and build and deploy the exe I had a quick look at your keygen I also have the keygen in my repo but I got all the values out of Ida to xor the values against so there is two ways to skip the check the easy way will just read plane text from the user.cod file if you want to make white space I'll confirm later but you just xor the xor values so they become 0 you can find the values in my repo in the main.cpp of the keygen but if you go the hardroute the xor happens twice before some checks are made on the value of chars at set ofsets that's where I'm looking now and what happens is because of the 2nd xor the values get changed again

**Comment by ghost** [comment]
Thanks, try my version out no keygen needed. USER.Cod can be edited to anything up to 100 characters😁😁

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by blkph0x** [reply]
![image](https://user-images.githubusercontent.com/26495065/217411927-480e3af7-ea8b-4609-b09b-37deea17f22b.png)
so in this, you can see the 1st jump if zero if we break on this and then jmp to paulo no user.cod file needed but if you break on the 2nd jump if not zero you will require a user.cod file to reach that code then if its there and 100 will go to the two calls to sub_ADD918 thats where the xor happens twice using hardcoded values to xor with then i assume after that is when it checks what the value of each char is if you notice on a licensed version the username callsign and such are spaced out the call sign can be used in some modes as well as location when we do it this way we are not conforming to the structure it requires so a full keygen is a bit harder so thats why i patched its so i can still play around with the xor operation so if you run the 1st xor and then jump you can see how the xor function is working step it twice and jump from the 2nd its evident a 2nd xor takes place have a few days off now so might dive back in and sort out the xor and then no patch required and if they change the key we can dump that to update the keygen

**Comment by ghost** [comment]
ive changed the the 3 xors and the 2 move ebp,(eax+edx)loops. It runs with whatever is put in USER.Cod as the registration

Kat🦄
________________________________

**Comment by ghost** [comment]
i couldnt work out where it was setting flags to keep saying not registered so copied your jumps to get it working . Had to do one more jnz in clock.exe. No keygen now required. Think its already being shared with 100 spaces put in USER.Cod 😂😂  Guess they just happy to gave it unlocked abs don't care about personalising a greeting . i gave you credit though for original ideas

Kat🦄
________________________________

**Reply by blkph0x** [reply]
you can just jump to paulo from here 
![image](https://user-images.githubusercontent.com/26495065/217417932-0cdba3f5-ab25-4d7b-ad13-01a31dd3987e.png)
 at the jnz loc_ADE2C8 patch that to a jmp to paulos block and the user.cod file can have anything in it. cheers :) i wana get a working keygen going without a patch that will make it harder for it to be patched ill keep you in the loop and if you figure any more out keep us updated

**Comment by ghost** [comment]
ok

Kat🦄
________________________________

**Comment by ghost** [comment]
i knocked out all his checks though, so there is no password code checking or character comparisons.
Kat🦄
________________________________

**Comment by ghost** [comment]
have you got his latest test version to keep keygen updated?    http://f6cte.free.fr/MULTIPSK_4_48_1_test_version_3.zip

Kat🦄
________________________________

**Comment by ghost** [comment]
Heres what i did...

==============================
MultiPSK 4.48 Cracking Notes:
===============================

Objective - change all XOR codes and move                 loops, which manipulate
characters from USER.COD with the XOR codes stored at:
=============================================================================

BREAKPOINTS   0052279E    RUN DEBUGGER BUT STOP AT ENTRY POINT NEEDED
              00522554    XOR DL,CL - First XOR TO BE CHANGED
              0052211C    MOV AL, [EDX+EAX-1] TO BE CHANGED
              00522825

=========================================================================================
MANIPULATION OF USER.COD CHARACTERS STEP 1:

      CHANGE 00522554 xor     dl, cl  --  32 D1 88 54 38 FF 43 46 80 FB 65 75 D5 33 C0 5A
          TO 00522554 xor     cl, dl  --  30 D1

      CODE:00522536 XOR_USERCODstep1:
      CODE:00522536 mov     eax, offset dword_CC98C0
      CODE:0052253B call    sub_405924
      CODE:00522540 mov     edi, ebx
      CODE:00522542 and     edi, 0FFh
      CODE:00522548 mov     edx, ds:dword_CC98C0
      CODE:0052254E mov     dl, [edx+edi-1]
      CODE:00522552 mov     cl, [esi]
      CODE:00522554 xor     dl, cl
      CODE:00522556 mov     [eax+edi-1], dl
      CODE:0052255A inc     ebx
      CODE:0052255B inc     esi
      CODE:0052255C cmp     bl, 65h ; 'e'
      CODE:0052255F jnz     short XOR_USERCODstep1

 NOTE:  GREEN CODE AT 00402BFC MAKES A SECOND COPY OF USER.COD IN MEMORY:
==========================================================================================
MANIPULATION OF USER.COD CHARACTERS STEP 2:

      CHANGE 0052211C MOVE [EDX],A1 --  A1 88 02 43 46 80 FB 65 75 DB 8B C7 8B 55 FC E8 25

            TO 0052211C XOR CL,DL     --  30 D1 (DOES NOTHING BUT FILL THE 2 BYTES NEEDED TO CHANGE)

      CODE:00522100 USERCODchangecharactersSTEP2:
      CODE:00522100 lea     eax, [ebp+var_4]
      CODE:00522103 call    sub_405924
      CODE:00522108 xor     edx, edx
      CODE:0052210A mov     dl, bl
      CODE:0052210C lea     eax, [eax+edx-1]
      CODE:00522110 push    eax
      CODE:00522111 xor     eax, eax
      CODE:00522113 mov     al, [esi]
      CODE:00522115 mov     edx, [edi]
      CODE:00522117 mov     al, [edx+eax-1]
      CODE:0052211B pop     edx
      CODE:0052211C mov     [edx], al
      CODE:0052211E inc     ebx
      CODE:0052211F inc     esi
      CODE:00522120 cmp     bl, 65h ; 'e'
      CODE:00522123 jnz     short USERCODchangecharactersSTEP2

 NOTE:  CODE AT 00402BFC IS THEN REVISTED AND MAKES A THIRD COPY OF USER.COD IN MEMORY:

      0A39F020  64 00 00 00 50 61 74 63  68 65 64 20 77 69 74 68  d...Patched with
      0A39F030  20 24 65 6C 6F 74 61 70  65 3A 29 20 74 6B 73 20   $elotape:) tks
      0A39F040  74 6F 20 42 6C 6B 50 68  30 78 20 6F 72 69 67 69  to BlkPh0x origi
      0A39F050  6E 61 6C 20 69 64 65 61  3A 29 29 20 20 20 23 23  nal idea:))   ##
      0A39F060  20 45 44 49 54 20 55 53  45 52 2E 43 4F 44 20 54   EDIT USER.COD T
      0A39F070  4F 20 59 4F 55 52 20 4F  57 4E 20 50 52 45 46 45  O YOUR OWN PREFE
      0A39F080  52 45 4E 43 45 20 23 23  8C 30 00 00 77 00 00 00  RENCE ##.0..w...
      0A39F090  02 00 00 00 64 00 00 00  50 61 74 63 68 65 64 20  ....d...Patched
      0A39F0A0  77 69 74 68 20 24 65 6C  6F 74 61 70 65 3A 29 20  with $elotape:)
      0A39F0B0  74 6B 73 20 74 6F 20 42  6C 6B 50 68 30 78 20 6F  tks to BlkPh0x o
      0A39F0C0  72 69 67 69 6E 61 6C 20  69 64 65 61 3A 29 29 20  riginal idea:))
      0A39F0D0  20 20 23 23 20 45 44 49  54 20 55 53 45 52 2E 43    ## EDIT USER.C
      0A39F0E0  4F 44 20 54 4F 20 59 4F  55 52 20 4F 57 4E 20 50  OD TO YOUR OWN P
      0A39F0F0  52 45 46 45 52 45 4E 43  45 20 23 23 00 00 00 00  REFERENCE ##....
      0A39F100  76 00 00 00 01 00 00 00  64 00 00 00 50 61 74 63  v.......d...Patc
      0A39F110  68 65 64 20 77 69 74 68  20 24 65 6C 6F 74 61 70  hed with $elotap
      0A39F120  65 3A 29 20 74 6B 73 20  74 6F 20 42 6C 6B 50 68  e:) tks to BlkPh
      0A39F130  30 78 20 6F 72 69 67 69  6E 61 6C 20 69 64 65 61  0x original idea
      0A39F140  3A 29 29 20 20 20 23 23  20 45 44 49 54 20 55 53  :))   ## EDIT US
      0A39F150  45 52 2E 43 4F 44 20 54  4F 20 59 4F 55 52 20 4F  ER.COD TO YOUR O
      0A39F160  57 4E 20 50 52 45 46 45  52 45 4E 43 45 20 23 23  WN PREFERENCE ##

================================================================================================
MANIPULATION OF USER.COD CHARACTERS STEP 3:


   CODE FROM STEP 2 IS REVISTED AND SO NOTHING NEEDS CHANGING AS ALREADY PATCHED
              ((0052211C MOVE [EDX],A1 ALREADY PATCHED TO XOR CL,DL))   :)

================================================================================================


      00522FD MOV [EAX+EDX], CL --  88 0C 10 43 80 FB 64 0F 85 D1 FE FF FF 8B C7 BA

         ALTER BYTES TO   NOPs  --  90 90 90



================================================================================================
FINALLY NEED TO CHANGE THE WAY PROGRAME RUNS TO USE THE CHANGES ABOVE SO NO CODE IS NEEDED, AND
USER.COD CAN CONTAIN ANY PERSONALISATION, GREETING OR JUST BLANK, BUT BE FULLY ACTIVATED :))

SO CHANGE THE JZ LOC 522A87 JNZ-- 0F 85 B8 02 00 00 E8 00 F9 FF FF E8 FB F8 FF FF

            ALTER BYTES TO  JNS -- 0F 89 27

            **to point to the "Paulo Santos ROUTINE" AT 005229F6 as found by BlkPh0x :)))**








________________________________

**Comment by blkph0x** [comment]
long days loo 
UPDATE:
so I have done a lot more reversing the serial not only does it xor the values it uses the same key as a decimal value to grab char at the offset of each decimal in the key so if you input the alphabet S being at the 19th offset a new string is built this happens twice once with our input then with the returned values of the 1st run of the call ADD918 then the 1st 4 values of the new string are placed into byte_634D56C byte_634D56E byte_634D56F for use with the call sub_ADDC90 byte_634D56D before we get really close to paulo so once we get to sub_ADDB94 it takes the new first 4 chars and does a little math will work that math out today I have already reversed the function that shifts the bytes around to make the new string then I xor the vales with our key and all values are pushing through until we hit this new calculator function but its only 4 values so will have that sorted today and hopefully soon a working keygeb

**Comment by ghost** [comment]
Glad to see you are progressing. I did come across a section where there is a user database decoded and compared! But could not see a common way to use the XOR coding. I have found out why my old way of using 00 (NOP) would not work 😂😂😂  it is now code 90 for NOP so i have NOPed out the relevant bytes to stop XOR and moving bytes around. Im not sure if test version 3 is the latest one he has issued or was included in the release of v4.48. But i applied my same routine to that test version and got the same activated results. Ive stopped messing with it now. Was trying to reverse DX Atlas but it is tricky as he has scrambled the code, and it unscrambles as it goes along. Attached are my completed notes on Multipsk. The XOR codes used cannot be found in a HEX editor to patch the programme directly, as was thinking of making a patch to use own XOR codes instead.  Keep me updated on keygen though
Kat
________________________________

**Reply by blkph0x** [reply]
![image](https://user-images.githubusercontent.com/26495065/218289438-da727ffd-3331-492a-90fb-8e64c744b872.png)
this is what happens so the order there of the bytes is SNKA so how that happens is it uses 
00CC99F0  13 01 5C 3F 43 38 26 2B  46 25 2C 05 10 2F 04 51  ..\?C8&+F%,../.Q
00CC9A00  0A 37 3A 09 19 2D 11 0F  49 32 31 33 24 15 64 08  .7:..-..I213$.d.
00CC9A10  57 35 63 4B 1D 3C 28 06  4D 30 54 0C 52 42 5A 56  W5cK.<(.M0T.RBZV
00CC9A20  23 3D 14 29 58 0E 3E 0D  1F 36 3B 53 0B 17 27 4E  #=.)X.>..6;S..'N
00CC9A30  4C 62 1B 39 1A 5E 03 07  44 5D 47 20 60 1C 21 1E  Lb.9.^..D]G `.!.
00CC9A40  5B 40 2E 50 22 2A 45 02  48 16 4F 59 18 61 55 5F  ***@***.***"*E.H.OY.aU_
00CC9A50  4A 41 34 12 64 00 00 00  00 00 00 00 4C 32 63 0F  JA4.d.......L2c.
so if we input abcdefghijklmnopqrstuvwxyz it will grab the dec value of the 1st char at 00CC99F0 so is 19 so is "s" 01 will be "a" 5C  92 so yeah any how thats how it works does that 4 times then will zoom on over to 00ADE069 start checking some other things it uses the 1st 4 values it moves to the start in this section  so when we have control over writing the serial before we xor with the xor list we can place these 4 important bytes at offsets 19 01 92 63
L BCDEFGHIJKLMNOPQR B TUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJ P 
LMNOPQRSTUVWXYZABCDEFGHIJKLM K OPQRSTUS
![image](https://user-images.githubusercontent.com/26495065/218290407-0ecb8618-cbbf-4b3c-a43f-32449840e416.png)
more to come

**Reply by blkph0x** [reply]
![image](https://user-images.githubusercontent.com/26495065/218290610-71caf4c2-823c-41d7-b25e-546aa8fb59d4.png)
this is what its doing with the 4 values at the ofsets i have shown

**Reply by blkph0x** [reply]
not sure how up with things you are but my methodology for reversing this is to first build the functions as they are then systematicily work backwards to work in a reverse order so if you check out  https://github.com/blkph0x/WIP-MultiPSK-keygen/blob/main/Multipsk_keygen/Multipsk_keygen.cpp updated with some more of the logic

**Comment by ghost** [comment]
I had a little play this afternoon, revisting the code area where it compares the registered users against the USER.COD data.
I managed to get programme running, but a completly different registered username. Then got a crash error!!
Heres my notes

MULTIPSK JUMPING THROUGH AND INTO THE CHECK REGISTERED USERS CODE:

CODE  AT 0522953 SETS A FLAG TO "C"  OR "P"   COMMERICAL AND PERSONAL???
(CURRENT CODE WILL SET IT TO P)

*****************************************************************
KEEPING FLAGS SET TO 1 TO ENABLE PATCHING THE WAY THROUGH, CHANGE:
*****************************************************************

NOTE MUST CHANGE ALL THE CODE SO THAT WE DO NOT JUMP TO TO 00522A87:

00522966 CMP A1,A1  84 C0 76 2D 88 45 E7 B3 01 33 C0 8A C3 8B 17 8A
CHANGE              90 90


005229EA  JB LOC 522A87 0F 82 97 00 00 00 80 7D FF 01 75 33 8D 45 EC 50
CHANGE                  90 90 90 90 90 90

*******************************************
MultiPSK Regisered Users decode at 0052343E
*******************************************


CODE:0052343E loc_52343E:
CODE:0052343E mov     edx, ebx
CODE:00523440 mov     eax, dword ptr ds:unk_CC9AC4
CODE:00523445 mov     al, [eax+ebx-1]
CODE:00523449 call    sub_522C44
CODE:0052344E push    eax
CODE:0052344F mov     eax, offset unk_CC9AC4
CODE:00523454 call    sub_405924
CODE:00523459 pop     edx
CODE:0052345A mov     [eax+ebx-1], dl
CODE:0052345E inc     ebx
CODE:0052345F cmp     ebx, 51h ; 'Q'
CODE:00523462 jnz     short loc_52343E

After this loop is completed EAX at 0A2CF0DC contains a registered user:

0A2AF1A0  60 00 00 00 63 00 00 00  01 00 00 00 50 00 00 00  `...c.......P...
0A2AF1B0  41 62 65 6C 20 44 61 73  69 6C 76 61 20 28 4B 45  Abel Dasilva (KE
0A2AF1C0  31 41 46 29 20 20 20 20  20 20 20 20 31 30 37 20  1AF)        107
0A2AF1D0  54 69 65 72 6E 61 6E 20  41 76 65 6E 75 65 20 20  Tiernan Avenue
0A2AF1E0  20 20 57 61 72 77 69 63  6B 2C 20 52 49 20 30 32    Warwick, RI 02
0A2AF1F0  38 38 36 20 20 20 20 20  20 20 20 20 20 55 53 41  886          USA


0A2EF1A0  60 00 00 00 63 00 00 00  01 00 00 00 50 00 00 00  `...c.......P...
0A2EF1B0  41 63 63 6F 6D 61 7A 7A  6F 20 47 69 61 6E 20 4D  Accomazzo Gian M
0A2EF1C0  61 72 63 6F 20 28 49 57  31 45 43 44 29 20 20 20  arco (IW1ECD)
0A2EF1D0  20 20 20 56 69 61 20 4D  61 72 63 6F 6E 69 20 32     Via Marconi 2
0A2EF1E0  34 20 20 20 20 31 35 30  34 30 20 47 72 61 76 61  4    15040 Grava
0A2EF1F0  20 28 41 4C 29 20 20 20  20 20 20 49 54 41 4C 59   (AL)      ITALY

0A2EF1A0  60 00 00 00 63 00 00 00  01 00 00 00 50 00 00 00  `...c.......P...
0A2EF1B0  41 64 61 6C 62 65 72 74  6F 20 4D 61 72 69 61 20  Adalberto Maria
0A2EF1C0  54 61 73 73 69 20 20 20  20 20 20 20 20 20 20 56  Tassi          V
0A2EF1D0  69 61 20 64 69 20 43 61  73 61 6C 20 53 65 6C 63  ia di Casal Selc
0A2EF1E0  65 20 33 35 30 20 20 30  30 31 36 36 20 52 6F 6D  e 350  00166 Rom
0A2EF1F0  61 20 20 20 20 20 20 20  20 20 20 49 54 41 4C 59  a          ITALY


BUT  AT CODE 00405847 COMPARES BETWEEN REGISTERED USER AND USER.COD CODE START:
MUST STOP ZERO FLAGS CAUSING JUMPS TO 4058A1

00405847 CHANGE 75 58 4A 74 15 8B 4E 04 8B 5F 04 39 D9 75 4B 83
       TO       90 90

00405854 CHANGE 75 4B 83 C6 08 83 C7 08 4A 75 E2 EB 06 83 C6 04
       TO       90 90

ETC ETC


THE KEY TO DECRYPTING REGISTERED USERS IS AT
00CC9AC0  00 00 00 00 00 00 00 00  0D 0F 2E 2D 16 49 2B 12  ...........-.I+.
00CC9AD0  34 39 04 25 4A 4B 13 22  0E 21 2C 23 35 2F 50 01  49.%JK.".!,#5/P.
00CC9AE0  1E 0C 08 18 42 3E 4C 3A  40 26 32 09 47 37 43 14  ....B>L:@&2.G7C.
00CC9AF0  4D 41 20 2A 1F 02 31 28  0A 4E 1A 33 03 15 05 44  MA *..1(.N.3...D
00CC9B00  3B 29 24 30 3C 4F 06 0B  17 46 19 27 45 48 3F 36  ;)$0<O...F.'EH?6
00CC9B10  11 10 1C 3D 1B 07 38 1D  00 00 00 00 00 00 00 00  ...=..8.........




________________________________
